apollo-api-core/src/main/kotlin/com/apollographql/apollo/api/CustomTypeAdapter.kt

```diff
@@ -0,0 +1,62 @@
+package com.apollographql.apollo.api
+
+/**
+ * Class represents the adapter for mapping GraphQL custom scalar types to Java objects.
+ *
+ * The GraphQL specification includes the following default scalar types: **String, Int, Float and Boolean**. But
+ * often use cases arise when support for custom scalar types is required and for those cases this class should be used
+ * to customize conversion. Here's an example of a type adapter for scalar type `java.util.Date`:
+ *
+ * ```
+ *    val adapter = object : CustomTypeAdapter<Date> {
+ *
+ *      override fun(value: CustomTypeValue<*>): Date {
+ *        try {
+ *          return ISO8601_DATE_FORMAT.parse(value);
+ *        } catch (ParseException e) {
+ *          throw new RuntimeException(e);
+ *        }
+ *      }
+ *
+ *      override fun encode(value: Date): CustomTypeValue {
+ *        return ISO8601_DATE_FORMAT.format(value);
+ *      }
+ *   }
+ * ```
+ *
+ * To use a custom type adapter with Apollo, you must first define the mapping in the build.gradle file:
+ * ```
+ *  apollo {
+ *    customTypeMapping['DateTime'] = "java.util.Date"
+ *  }
+ * ```
+ *
+ * The lines above tell the compiler which type to use while generating the classes. Once this is done, register the type
+ * adapter with `com.apollographql.apollo.ApolloClient.Builder`:
+ *
+ * ```
+ *  val builder = ApolloClient.builder()
+ *    .addCustomTypeAdapter(CustomType.DATE, dateCustomTypeAdapter)
+ * ```
+ *
+ * In the example code above, the `CustomType.DATE` is the class generated by the compiler.
+ */
+interface CustomTypeAdapter<T> {
+
+  /**
+   * De-serializes the value to the custom scalar type. Usually used in parsing the GraphQL response.
```

Comment:
In KDoc, Jetbrains usually make the params and return the part of the sentence and do not include separate `@param / @return` section. Something like

```
De-serializes the [value] and return the custom scalar type [T]. Usually used in parsing the GraphQL response
```